Obscured input fields are not obscured in the keyboard on visionOS.

megangardner opened the pull request:
#### 4cde644bbab2419c9590436ba32925208bb90aa6
<pre>
Obscured input fields are not obscured in the keyboard on visionOS.
<a href="https://bugs.webkit.org/show_bug.cgi?id=260838">https://bugs.webkit.org/show_bug.cgi?id=260838</a>
rdar://113924534

Reviewed by Wenson Hsieh.

We need to look at the flag that is set on input fields that indicates that
they are obscured and use that information to determine if we should
emit all the of text. We started emitting text in
<a href="https://github.com/WebKit/WebKit/commit/e1c274e43bdde86eea8fdd34ccc7789fdc279511">https://github.com/WebKit/WebKit/commit/e1c274e43bdde86eea8fdd34ccc7789fdc279511</a>
so that we would show password in the keyboard field on visionOS,
but if the isAutoFilledAndObscured flag is set, we do not want to display
that information in the keyboard. We look at the start and end of the selected range,
and if either one is in an obscured input field, we do not emit all text.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDocumentEditingContext):

Canonical link: <a href="https://commits.webkit.org/267436@main">https://commits.webkit.org/267436@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/239482c2e6204e5b77c5c5465d8db0b8e80c0596

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16545 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16868 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17309 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18327 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15517 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20098 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17009 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17856 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16743 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17146 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14308 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19109 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14390 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21788 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15378 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15158 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19474 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15758 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13378 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14958 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3968 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19327 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15583 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -4944,12 +4944,28 @@ static VisiblePositionRange constrainRangeToSelection(const VisiblePositionRange
-    auto makeString = [] (const VisiblePosition& start, const VisiblePosition& end) -> AttributedString {
+    auto isTextObscured = [] (const VisiblePosition& visiblePosition) -> bool {
+        if (RefPtr textControl = enclosingTextFormControl(visiblePosition.deepEquivalent())) {
+            if (is<HTMLInputElement>(textControl)) {
```
```suggestion
            if (auto *input = dynamicDowncast<HTMLInputElement(textControl); input.isAutoFilledAndObscured()) {
            return true
            }
```

I think using `dynamicDowncast` might be a bit cleaner

pxlcoder replied:
`auto* ` rather than `auto *` from Richard's comment, since it's a C++ object.

whsieh commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -4944,12 +4944,28 @@ static VisiblePositionRange constrainRangeToSelection(const VisiblePositionRange
         // FIXME: This should return editing-offset-compatible attributed strings if that option is requested.
-        return WebCore::AttributedString::fromNSAttributedString(adoptNS([[NSAttributedString alloc] initWithString:WebCore::plainTextReplacingNoBreakSpace(*range, TextIteratorBehavior::EmitsOriginalText)]));
+
+        auto isObscured = isTextObscured(start) || isTextObscured(end);
```
Nit - I think just checking only the start should be sufficient — `enclosingTextFormControl` should always return the same shadow host in both cases.

You could then inline the helper lambda above into this (though keeping them separate is okay too)

aprotyas commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -4944,12 +4944,28 @@ static VisiblePositionRange constrainRangeToSelection(const VisiblePositionRange
     }
 
-    auto makeString = [] (const VisiblePosition& start, const VisiblePosition& end) -> AttributedString {
+    auto isTextObscured = [] (const VisiblePosition& visiblePosition) -> bool {
```
nit: I think we can do without the trailing return type on this.

aprotyas commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -4944,12 +4944,28 @@ static VisiblePositionRange constrainRangeToSelection(const VisiblePositionRange
+        return false;
+    };
+
+    auto makeString = [&isTextObscured] (const VisiblePosition& start, const VisiblePosition& end) -> AttributedString {
```
Ditto comment about the return type.

megangardner committed:
Obscured input fields are not obscured in the keyboard on visionOS.
https://bugs.webkit.org/show_bug.cgi?id=260838
rdar://113924534

Reviewed by Wenson Hsieh.

We need to look at the flag that is set on input fields that indicates that
they are obscured and use that information to determine if we should
emit all the of text. We started emitting text in
https://github.com/WebKit/WebKit/commit/e1c274e43bdde86eea8fdd34ccc7789fdc279511
so that we would show password in the keyboard field on visionOS,
but if the isAutoFilledAndObscured flag is set, we do not want to display
that information in the keyboard. We look at the start and end of the selected range,
and if either one is in an obscured input field, we do not emit all text.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDocumentEditingContext):

Canonical link: https://commits.webkit.org/267436@main
